Chaincase.Common/Services/AesThenHmac.cs

```diff
@@ -0,0 +1,402 @@
+
+            byte[] cipherText;
+            byte[] iv;
+
+            using (var aes = new AesManaged
+            {
+                KeySize = KeyBitSize,
+                BlockSize = BlockBitSize,
+                Mode = CipherMode.CBC,
+                Padding = PaddingMode.PKCS7
+            })
+            {
+
+                //Use random IV
+                aes.GenerateIV();
+                iv = aes.IV;
+
+                using (var encrypter = aes.CreateEncryptor(cryptKey, iv))
+                using (var cipherStream = new MemoryStream())
+                {
+                    using (var cryptoStream = new CryptoStream(cipherStream, encrypter, CryptoStreamMode.Write))
+                    using (var binaryWriter = new BinaryWriter(cryptoStream))
+                    {
+                        //Encrypt Data
+                        binaryWriter.Write(secretMessage);
+                    }
+
+                    cipherText = cipherStream.ToArray();
+                }
+
+            }
+
+            //Assemble encrypted message and add authentication
+            using (var hmac = new HMACSHA256(authKey))
+            using (var encryptedStream = new MemoryStream())
+            {
+                using (var binaryWriter = new BinaryWriter(encryptedStream))
+                {
+                    //Prepend non-secret payload if any
+                    binaryWriter.Write(nonSecretPayload);
+                    //Prepend IV
+                    binaryWriter.Write(iv);
+                    //Write Ciphertext
+                    binaryWriter.Write(cipherText);
+                    binaryWriter.Flush();
+
+                    //Authenticate all data
+                    var tag = hmac.ComputeHash(encryptedStream.ToArray());
+                    //Postpend tag
+                    binaryWriter.Write(tag);
+                }
+                return encryptedStream.ToArray();
+            }
+
+        }
+
+        public static byte[] Decrypt(byte[] encryptedMessage, byte[] cryptKey, byte[] authKey, int nonSecretPayloadLength = 0)
+        {
+
+            //Basic Usage Error Checks
+            if (cryptKey == null || cryptKey.Length != KeyBitSize / 8)
+                throw new ArgumentException(String.Format("CryptKey needs to be {0} bit!", KeyBitSize), "cryptKey");
+
+            if (authKey == null || authKey.Length != KeyBitSize / 8)
+                throw new ArgumentException(String.Format("AuthKey needs to be {0} bit!", KeyBitSize), "authKey");
+
+            if (encryptedMessage == null || encryptedMessage.Length == 0)
+                throw new ArgumentException("Encrypted Message Required!", "encryptedMessage");
+
+            using (var hmac = new HMACSHA256(authKey))
+            {
+                var sentTag = new byte[hmac.HashSize / 8];
+                //Calculate Tag
+                var calcTag = hmac.ComputeHash(encryptedMessage, 0, encryptedMessage.Length - sentTag.Length);
```

Comment:
probably not really an issue, but this will throw if encryptedMessage is malformed and < the size of the IV

Reply:
If we implement the check and encryptedMessage.length <= 1, shouldn't we also throw? If so, isn't that error good?